### PR TITLE
Add many error handling

### PR DESCRIPTION
**Pull request by SuperKenVery**

### CODE REVIEW

**Comment by SuperKenVery** [comment]
Also, a new `AsyncChatGPT`

**Comment by SuperKenVery** [comment]
Ah sorry, wait for me to debug please... So many bugs currently :(

**Comment by SuperKenVery** [comment]
Please have a look! :pray: :pray:  This might seem a lot of changes (indeed) but I think they largely improve this thing!

Summary: 
- Added a `AsyncChatGPT`
- Added many error handling
- Improved the `refresh_session` function: Navigate the browser to `/api/session/auth` rather than making a plain XHR request, so that we can pass the browser check. 
- Use a single browser and page across different instances of `ChatGPT` (but still different conversations), for less risk of getting `Access denied` or something like that. 

This results in a much more stable behavior and more up time, according to my feelings when I use this library to build a [qq bot](https://gitee.com/xyhken/nju-verification-kun.git). I believe this can help many people who also wants to build a bot.

File: chatgpt_wrapper/asyncChatGPT.py [hunk 1]
```diff
@@ -0,0 +1,669 @@
+""" AsyncChatGPT class
```
Comment (thehunmonkgroup) [inline chatgpt_wrapper/asyncChatGPT.py:1]: This class creates a ton of code duplication, and would make maintenance far worse and more error prone. What is the purpose/advantage of having a completely separate async class?

Even if it is a good idea, it is definitely out of scope for this PR, and should be put in its own PR.

File: chatgpt_wrapper/chatgpt.py [hunk 2]
```diff
@@ -1,26 +1,29 @@
+""" AsyncChatGPT class
```
Comment (thehunmonkgroup) [inline chatgpt_wrapper/chatgpt.py:1]: Inaccurate doc.

File: chatgpt_wrapper/chatgpt.py [hunk 3]
```diff
@@ -35,40 +38,67 @@ class ChatGPT:
...
     session_div_id = "chatgpt-wrapper-session-data"
+    error_div_id = "chatgpt-wrapper-error-data"
+    lock=threading.Lock()
+    browser=None
+    page=None
+    enabled=True
+    session={}
 
     def __init__(self, headless: bool = True, browser="firefox", model="default", timeout=60, debug_log=None, proxy: Optional[ProxySettings] = None):
+        """Create an AsyncChatGPT asynchronously
```
Comment (thehunmonkgroup) [inline chatgpt_wrapper/chatgpt.py:49]: Inaccurate doc.

File: chatgpt_wrapper/chatgpt.py [hunk 4]
```diff
@@ -35,40 +38,67 @@ class ChatGPT:
...
-                headless=headless,
-                proxy=proxy,
-            )
 
-        if len(self.browser.pages) > 0:
-            self.page = self.browser.pages[0]
-        else:
-            self.page = self.browser.new_page()
-        self._start_browser()
+        if ChatGPT.browser==None:
```
Comment (thehunmonkgroup) [inline chatgpt_wrapper/chatgpt.py:68]: What are you using `ChatGPT` here instead of `self`? Unless I'm misunderstanding, it would limit the class to only being able to work with one browser across multiple instances of the class.

I would certainly hope it's possible in Playwright to spin up multiple browser instances, and I see no reason to limit a user of the class to only being able to use one.

File: chatgpt_wrapper/chatgpt.py [hunk 5]
```diff
@@ -88,48 +118,71 @@ def _set_logging(self, debug_log):
...
     def _start_browser(self):
-        self.page.goto("https://chat.openai.com/")
+        self.log.info("Loading ChatGPT webpage...")
+        ChatGPT.page.goto("https://chat.openai.com/")
+        try:
+            ChatGPT.page.wait_for_url("/chat")
+        except Exception:
+            errors=ChatGPT.page.query_selector_all("div.cf-error-details")
+            if len(errors)>0 and 'Access denied' in errors[0].inner_html():
+                self.log.error("Got Access denied. Self-disabling for 10 minutes. ")
```
Comment (thehunmonkgroup) [inline chatgpt_wrapper/chatgpt.py:128]: Not loving the hard-code self disabling thing, what is the purpose of this code? Seems as likely to accidentally lock a user out as to provide benefit.

File: chatgpt_wrapper/chatgpt.py [hunk 6]
```diff
@@ -333,48 +411,222 @@ def ask_stream(self, prompt: str):
...
+            self.log.error("Timeout when asking ChatGPT. Retrying...")
+            self.refresh_session()
+            for chunk in self.ask_stream(prompt=prompt,remaining_retry=remaining_retry-1):
+                yield chunk
+            return
+
+        self._cleanup_divs()
+
+    def ask_stream_clicking(self, prompt: str, remaining_retry=2):
+        """Ask ChatGPT by clicking the buttons on the website. Response is streamed.
```
Comment (thehunmonkgroup) [inline chatgpt_wrapper/chatgpt.py:446]: What is the purpose of this function versus the other streaming function?

File: chatgpt_wrapper/chatgpt.py [hunk 7]
```diff
@@ -333,48 +411,222 @@ def ask_stream(self, prompt: str):
...
+        return self.ask(message,remaining_retry=remaining_retry-1)
 
     def new_conversation(self):
         self.parent_message_id = str(uuid.uuid4())
         self.conversation_id = None
         self.conversation_title_set = None
+
+
+
+def main():
```
Comment (thehunmonkgroup) [inline chatgpt_wrapper/chatgpt.py:612]: Why is this in here? We already have code to interact at shell.

File: chatgpt_wrapper/chatgpt.py [hunk 8]
```diff
@@ -1,26 +1,29 @@
...
+import error
 
 RENDER_MODELS = {
     "default": "text-davinci-002-render-sha",
     "legacy-paid": "text-davinci-002-render-paid",
     "legacy-free": "text-davinci-002-render"
 }
 
-DEFAULT_CONSOLE_LOG_LEVEL = logging.ERROR
+DEFAULT_CONSOLE_LOG_LEVEL = logging.DEBUG
```
Comment (thehunmonkgroup) [inline chatgpt_wrapper/chatgpt.py:26]: Can't do this, the console level needs to remain at `ERROR`, the user can pass the `--debug-log` argument to get more detailed logs written to a file.

**Review by thehunmonkgroup — CHANGES_REQUESTED** [review]
Overall, this PR is trying to do too much, which makes it very difficult to review and suggest feedback. The feedback I offered is a start, but I'm sure I would have more if there was:

* More discussion about these changes prior to the PR
* Multiple, focus PRs to incrementally improve/change the code

**Comment by mmabrouk** [comment]
Hey @SuperKen,

Firstly, thank you for submitting the PR. After reading your comment, it seems that you have resolved many of the issues we have been experiencing lately. We have been struggling with refreshing sessions for the past few weeks, and introducing async is a much-needed feature that we still don't have.

However, I don't think we can merge the PR in its current form. The reason being, firstly, it introduces many changes at the same time, and secondly, it adds unnecessary redundancies and complexities.

My proposal is that you create a PR that only includes the session refresh feature, and then we can go from there. What do you think?

File: chatgpt_wrapper/asyncChatGPT.py (hunk 1 above)
Comment (SuperKenVery) [reply chatgpt_wrapper/asyncChatGPT.py:1]: The main problem I encountered is using an async generator in sync context. 

Here are what I have found:

I can't get [this](https://stackoverflow.com/questions/63587660/yielding-asyncio-generator-data-back-from-event-loop-possible/63595496#63595496) to run because I can't handle the `loop` argument. If I get a loop from asyncio it says there's currently no loop. If I create one it says the future I'm `awaiting` is in another loop. I really have little experience in async programming, so teach me if possible :smile: 

I also found another simpler solution, but that only returns all of the generated values when it has finished generating, which makes ask_stream meaningless. 

> Even if it is a good idea, it is definitely out of scope for this PR, and should be put in its own PR.

I'm sorry for the PR title. I'm not very familiar with github, pr and regular workflows. I was wondering that a PR can have only one title and whenever I submit new commits they will just add to this PR. I will create new branches for this :D

File: chatgpt_wrapper/chatgpt.py (hunk 4 above)
Comment (SuperKenVery) [reply chatgpt_wrapper/chatgpt.py:68]: With only one browser and one tab (page), I get less risk of getting banned by OpenAI. (For example, the `Access denied` page or something). Although I'm not totally sure about this (because I made this change along with many others), but I did see more up time of my bot. 

And what's the benefit of having multiple browser instances? As long as we specify different `conversation id`s, we're having different conversations.

File: chatgpt_wrapper/chatgpt.py (hunk 5 above)
Comment (SuperKenVery) [reply chatgpt_wrapper/chatgpt.py:128]: When `OpenAI` gives us an `Access denied` page, I'm afraid that continue to access (frequently) would prolong the denying time. 

This might need further investigating and if I'm wrong please tell me. 

(I don't have ChatGPT Plus)

File: chatgpt_wrapper/chatgpt.py (hunk 6 above)
Comment (SuperKenVery) [reply chatgpt_wrapper/chatgpt.py:446]: It asks ChatGPT by clicking the button on the website. 
- I doubted if the `text-davinci-002-render-sha` was actually the one used in ChatGPT. Through the `network` pane of browser developer tools I've seen different models (although also this one). Therefore, to ensure a total ChatGPT experience I wrote this. 
- I hoped it to be more stable than the XHR way, but it turns out it didn't help. 

However, as you're adding new model options and `ask_stream_clicking` hasn't implemented conversation choosing, I think I can remove it now.

File: chatgpt_wrapper/chatgpt.py (hunk 7 above)
Comment (SuperKenVery) [reply chatgpt_wrapper/chatgpt.py:612]: Ah. It was used as more or less a unit test. Teach me if it doesn't belong here.

File: chatgpt_wrapper/asyncChatGPT.py (hunk 1 above)
Comment (SuperKenVery) [reply chatgpt_wrapper/asyncChatGPT.py:1]: If this problem is solved I would like to preserve the async one and provide a sync wrapper over that.

File: chatgpt_wrapper/chatgpt.py (hunk 7 above)
Comment (SuperKenVery) [reply chatgpt_wrapper/chatgpt.py:488]: Or I should remove this. 

Actually I had some hard time figuring out how to test this code. Now I found that I should use `pip install -e`. lol

:laughing:

**Comment by SuperKenVery** [comment]
I think it would be too troubling and error-prone to seperate async changes and error handling changes, because error handling codes are everywhere. 

I think it might be better if we first merge the error handling part, then I PR the async part and we merge that. 

What do you think?

**Comment by thehunmonkgroup** [comment]
Yes, breaking it down into those two steps would certainly make it easier to review and approve the PRs, let's do that.

File: chatgpt_wrapper/chatgpt.py (hunk 6 above)
Comment (SuperKenVery) [reply chatgpt_wrapper/chatgpt.py]: Removed.

File: chatgpt_wrapper/chatgpt.py (hunk 7 above)
Comment (SuperKenVery) [reply chatgpt_wrapper/chatgpt.py:512]: Removed.

File: chatgpt_wrapper/chatgpt.py (hunk 8 above)
Comment (SuperKenVery) [reply chatgpt_wrapper/chatgpt.py]: Reverted.

**Comment by thehunmonkgroup** [comment]
Quick note that https://github.com/mmabrouk/chatgpt-wrapper/pull/169 ended up migrating Playwright to async, so that work is done for you!

**Comment by SuperKenVery** [comment]
Ohh that’s great! But… is it good to remove all the sync interfaces? Should we make a wrapper? 在 2023年2月28日，12:25，thehunmonkgroup ***@***.***> 写道：﻿
Quick note that #169 ended up migrating Playwright to async, so that work is done for you!

**Comment by thehunmonkgroup** [comment]
You know, that might be a good idea, I didn't do that yet, and ended up having to convert some of the GPTShell methods as well. I don't have a lot of experience with async programming in Python, and only did this work because it seemed necessary as part of the Prompt Toolkit migration.

**Comment by thehunmonkgroup** [comment]
After a night of sleep, it's now clear that my initial async implementation has some, well, flaws ;)

I'm now convinced that we need the sync wrapper you suggested, and I'll work on cleaning that up in a bit.

**Comment by SuperKenVery** [comment]
I think I might not spend time on this anymore... if nobody else is working on this or you want please just close this pr... 

Really had a great time cooperating with you!

**Comment by thehunmonkgroup** [comment]
Thanks for all your help on this project so far!
